### PR TITLE
Login Feature Initial Commit

### PR DESCRIPTION
**Pull request by CrowleyRajapakse**
## Purpose

- Implementing Login Functionality with authorization_code grant.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/constants/SQLConstants.java [hunk 1]
```diff
@@ -3093,4 +3093,19 @@
         public static final String CERTIFICATE_COUNT_QUERY = "SELECT COUNT(*) AS count FROM AM_API_CLIENT_CERTIFICATE " +
                 "WHERE TENANT_ID=? AND REMOVED=?";
     }
+
+    public static class SystemApplicationConstants {
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/constants/SQLConstants.java]: Class comments

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SystemApplicationDAO.java [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+                log.debug("Error while retrieving database information. ", e);
+            }
+            handleException("Error retrieving Database information", e);
+        } finally {
+            APIMgtDBUtil.closeAllConnections(null, connection, resultSet);
+        }
+        return isExists;
+    }
+
+    public boolean addApplicationKey(String appName, String consumerKey, String consumerSecret)
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SystemApplicationDAO.java]: Method comment

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SystemApplicationDAO.java [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+            }
+            handleException("Error while persisting client credentials to SYSTEM_APPS table ", e);
+        } finally {
+            APIMgtDBUtil.setAutoCommit(connection, initialAutoCommit);
+            APIMgtDBUtil.closeAllConnections(preparedStatement, connection, null);
+        }
+        return result;
+    }
+
+    public SystemApplicationDTO getClientCredentialsForApplication(String appName) throws APIMgtDAOException {
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SystemApplicationDAO.java]: Method comment

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SystemApplicationDAO.java [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+            }
+            handleException("Error while retrieving client credentials for application: " + appName, e);
+        } finally {
+            APIMgtDBUtil.setAutoCommit(connection, initialAutoCommit);
+            APIMgtDBUtil.closeAllConnections(preparedStatement, connection, resultSet);
+        }
+        return systemApplicationDTO;
+    }
+
+    public boolean removeConsumerKeyForApplication(String appName) throws APIMgtDAOException {
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SystemApplicationDAO.java]: Method comment

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SystemApplicationDAO.java [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+        try {
+            connection = APIMgtDBUtil.getConnection();
+            initialAutoCommit = connection.getAutoCommit();
+            connection.setAutoCommit(false);
+            preparedStatement = connection.prepareStatement(deleteApplicationKeyQuery);
+            preparedStatement.setString(1, appName);
+            result = preparedStatement.executeUpdate() == 1;
+            connection.commit();
+            connection.setAutoCommit(initialAutoCommit);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SystemApplicationDAO.java]: Set this in finally block

Comment (malinthaprasan) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SystemApplicationDAO.java]: Use APIMgtDBUtil.setAutoCommit method

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SystemApplicationDAO.java [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+        try {
+            connection = APIMgtDBUtil.getConnection();
+            initialAutoCommit = connection.getAutoCommit();
+            connection.setAutoCommit(false);
+            preparedStatement = connection.prepareStatement(checkClientCredentialsExistsQuery);
+            preparedStatement.setString(1, appName);
+            result = preparedStatement.executeUpdate() == 1;
+            connection.commit();
+            connection.setAutoCommit(initialAutoCommit);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SystemApplicationDAO.java]: Same here

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dto/SystemApplicationDTO.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.apimgt.impl.dto;
+
+import java.io.Serializable;
+import java.time.LocalDateTime;
+
+public class SystemApplicationDTO implements Serializable {
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dto/SystemApplicationDTO.java]: Class comments

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher-new/jaggery.conf [hunk 8]
```diff
@@ -15,8 +15,12 @@
             "path": "/services/routing/static.jag"
         },
         {
-            "url":"/services/*",
-            "path": "/services/login/app.jag"
+            "url":"/services/configs/config",
```
Comment (malinthaprasan) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher-new/jaggery.conf]: "url":"/services/config",

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher-new/jaggery.conf [hunk 9]
```diff
@@ -15,8 +15,12 @@
...
+            "path": "/services/configs/config.jag"
+        },
+        {
+            "url":"/services/configs/token",
```
Comment (malinthaprasan) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher-new/jaggery.conf]: "url":"/services/token",

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher-new/services/configs/config.jag [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    var clientSecret = result.data.clientSecret;
+    var scopes = "apim:api_view apim:api_create apim:api_publish apim:tier_view apim:tier_manage apim:subscription_view apim:subscription_block apim:mediation_policy_view apim:api_workflow openid";
+    var IDPHost = "localhost:9443";
+
+    var systemApplicationDAO = org.wso2.carbon.apimgt.impl.dao.SystemApplicationDAO;
+    var systemApplicationDAO = new SystemApplicationDAO();
+    var addApplicationKey = systemApplicationDAO.addApplicationKey("admin_publisher", clientId, clientSecret);
+    
+
+    var authorize_endpoint = "https://localhost:8243/authorize";
```
Comment (malinthaprasan) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher-new/services/configs/config.jag]: Hardcoded due to the initial implementation

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher-new/services/configs/config.jag [hunk 11]
```diff
@@ -0,0 +1,37 @@
+<%
+    var log = new Log("Services login DCR request");
+
+    var dcr_url = 'https://localhost:9443/client-registration/v0.14/register';
```
Comment (malinthaprasan) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher-new/services/configs/config.jag]: Hardcoded due to the initial implementation

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher-new/services/configs/token.jag [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    var systemApplicationDAO = new SystemApplicationDAO();
+    var systemApplicationDTO = systemApplicationDAO.getClientCredentialsForApplication("admin_publisher");
+
+    var clientId = systemApplicationDTO.getConsumerKey();
+    var clientSecret = systemApplicationDTO.getConsumerSecret();
+    var Base64 = org.apache.axiom.om.util.Base64;
+    var String = Packages.java.lang.String;
+    var base64encoded = Base64.encode ( new String (clientId + ":" + clientSecret).getBytes());
+    
+    var tokenEndpoint = "https://localhost:8243/token";
```
Comment (malinthaprasan) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher-new/services/configs/token.jag]: Hardcoded due to the initial implementation

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/constants/SQLConstants.java (hunk 1 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/constants/SQLConstants.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SystemApplicationDAO.java (hunk 2 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SystemApplicationDAO.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SystemApplicationDAO.java (hunk 3 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SystemApplicationDAO.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SystemApplicationDAO.java (hunk 4 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SystemApplicationDAO.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SystemApplicationDAO.java (hunk 5 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SystemApplicationDAO.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SystemApplicationDAO.java (hunk 6 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SystemApplicationDAO.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dto/SystemApplicationDTO.java (hunk 7 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dto/SystemApplicationDTO.java]: resolved

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher-new/jaggery.conf (hunk 8 above)
Comment (CrowleyRajapakse) [reply features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher-new/jaggery.conf]: resolved

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher-new/jaggery.conf (hunk 9 above)
Comment (CrowleyRajapakse) [reply features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher-new/jaggery.conf]: resolved

**Pull request merged**